File: kafka-repository/src/main/java/org/suggs/sandbox/eventsourcing/kafka/KafkaProducerBuilder.java
package org.suggs.sandbox.eventsourcing.kafka;

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.Producer;
import org.apache.kafka.clients.producer.ProducerConfig;

import java.util.Properties;

import static org.apache.kafka.clients.producer.ProducerConfig.*;

public class KafkaProducerBuilder {

    private Properties properties;

    private KafkaProducerBuilder() {
        properties = new Properties();
    }

    public static KafkaProducerBuilder aKafkaProducer() {
        return new KafkaProducerBuilder();
    }

    public KafkaProducerBuilder connectedTo(String kafkaUrl) {
        properties.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaUrl);
        return this;
    }

    public KafkaProducerBuilder knownAs(String clientName) {
        properties.put(ProducerConfig.CLIENT_ID_CONFIG, clientName);
        return this;
    }

    public KafkaProducerBuilder usingKeySerializer(String serializerName) {
        properties.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, serializerName);
        return this;
    }


    public KafkaProducerBuilder usingValueSerializer(String serializerName) {
        properties.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, serializerName);
        return this;
    }


    public <K, V> Producer<K, V> build() {
        return new KafkaProducer<>(properties);
    }

}

File: kafka-repository/src/main/kotlin/org/suggs/sandbox/eventsourcing/kafka/KafkaProducerBuilder.kt
package org.suggs.sandbox.eventsourcing.kafka

import org.apache.kafka.clients.producer.KafkaProducer
import org.apache.kafka.clients.producer.Producer
import org.apache.kafka.clients.producer.ProducerConfig
import java.util.*

class KafkaProducerBuilder {

    private val properties = Properties()

    companion object {
        fun aKafkaProducer() = KafkaProducerBuilder()
    }

    fun connectedTo(kafkaUrl: String) = addPropertiesAndReturn(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaUrl)
    fun knownAs(clientName: String) = addPropertiesAndReturn(ProducerConfig.CLIENT_ID_CONFIG, clientName)
    fun usingKeySerializer(serializerName: String?) = addPropertiesAndReturn(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, serializerName)
    fun usingValueSerializer(serializerName: String?) = addPropertiesAndReturn(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, serializerName)

    fun <K, V> build(): Producer<K, V> {
        return KafkaProducer(properties)
    }

    private fun addPropertiesAndReturn(aKey: String, aValue: String?): KafkaProducerBuilder {
        properties[aKey] = aValue
        return this
    }
}